frontend/js/src/common/brainzplayer/BrainzPlayer.tsx

```diff
@@ -487,82 +529,197 @@ export default class BrainzPlayer extends React.Component<
       !isListenFromDatasource(listen, datasource) &&
       !datasource.canSearchAndPlayTracks()
     ) {
-      this.playListen(listen, datasourceIndex + 1);
+      playListen(listen, nextListenIndex, datasourceIndex + 1);
       return;
     }
-    this.stopOtherBrainzPlayers();
-    this.setState({ currentDataSourceIndex: selectedDatasourceIndex }, () => {
+    stopOtherBrainzPlayers();
+    dispatch({ currentDataSourceIndex: selectedDatasourceIndex }, async () => {
+      while (
+        currentListenRef.current !== listen ||
+        currentDataSourceIndexRef.current !== selectedDatasourceIndex
+      ) {
+        // eslint-disable-next-line no-await-in-loop, no-promise-executor-return
+        await new Promise((resolve) => setTimeout(resolve, 100));
+      }
```

Comment:
So, here we'll have to wait for the `currentDataSourceIndex` to update, otherwise the player is not activated, and does not play the track.
https://github.com/metabrainz/listenbrainz-server/blob/566d5433b46079d3d2621fb9a735cb46a54e5094/frontend/js/src/common/brainzplayer/BrainzPlayer.tsx#L987-L991

Reply:
I get that part, I guess the part I was missing is that if we were to move this logic in BrainzPlayerContext, how would we access the value of currentDataSourceIndex?

mayyyybe we could pass the `currentDataSourceIndexRef` in the dispatch payload, and still do that same check with the `currentDataSourceIndex` prop we passed? At that point it might just make it all more complicated for no reason.


I feel like this mechanism is not really adapted to our new use of the BPContext.
Let's leave it like this for now, and maybe revisit how we structure things at a later date.